File: sdks/typescript/src/event_identity_values.ts
import { randomUUID } from "node:crypto";

function nonEmpty(value: string, label: string): string {
  if (typeof value !== "string" || value.trim().length === 0) {
    throw new TypeError(`${label} must not be empty`);
  }
  return value;
}

/** Validated event stream identity values. */
export class StreamId {
  readonly value: string;

  private constructor(value: string) {
    this.value = nonEmpty(value, "event stream_id");
    Object.freeze(this);
  }

  static new(value: string): StreamId {
    return new StreamId(value);
  }

  asStr(): string {
    return this.value;
  }

  toString(): string {
    return this.value;
  }
}

/** Validated event identity values. */
export class EventId {
  readonly value: string;

  private constructor(value: string) {
    this.value = nonEmpty(value, "event_id");
    Object.freeze(this);
  }

  static new(value: string): EventId {
    return new EventId(value);
  }

  asStr(): string {
    return this.value;
  }

  toString(): string {
    return this.value;
  }
}

export type EventRuntimeContext = {
  conversation_id?: string | null;
  run_id?: string | null;
  turn_id?: string | null;
  message_id?: string | null;
  execution_id?: string | null;
  parent_event_id?: string | null;
  subagent_lineage?: { parent_event_id?: string | null } | null;
};

/** Immutable transport identity projection without owning a live stream. */
export class EventIdentity {
  readonly streamId: StreamId;
  readonly conversationId: string | undefined;
  readonly runId: string | undefined;
  readonly turnId: string;
  readonly messageId: string | undefined;
  readonly executionId: string | undefined;
  readonly parentEventId: string | undefined;

  private constructor(fields: {
    streamId: StreamId;
    conversationId?: string;
    runId?: string;
    turnId: string;
    messageId?: string;
    executionId?: string;
    parentEventId?: string;
  }) {
    this.streamId = fields.streamId;
    this.conversationId = fields.conversationId;
    this.runId = fields.runId;
    this.turnId = nonEmpty(fields.turnId, "turn_id");
    this.messageId = fields.messageId;
    this.executionId = fields.executionId;
    this.parentEventId = fields.parentEventId;
    this.validate();
    Object.freeze(this);
  }

  static new(streamId: string, turnId: string): EventIdentity {
    return new EventIdentity({ streamId: StreamId.new(streamId), turnId });
  }

  static forRun(runId: string): EventIdentity {
    const id = nonEmpty(runId, "run_id");
    return new EventIdentity({
      streamId: StreamId.new(randomUUID()),
      runId: id,
      turnId: id,
      executionId: id,
    });
  }

  static forChat(
    conversationId: string | undefined,
    turnId: string,
    messageId: string,
    runId?: string,
  ): EventIdentity {
    return new EventIdentity({
      streamId: StreamId.new(randomUUID()),
      conversationId: conversationId === undefined ? undefined : nonEmpty(conversationId, "conversation_id"),
      runId: runId === undefined ? undefined : nonEmpty(runId, "run_id"),
      turnId,
      messageId: nonEmpty(messageId, "message_id"),
    });
  }

  static fromInvocation(invocation: { runtime?: EventRuntimeContext | null } | null | undefined): EventIdentity {
    return EventIdentity.fromRuntimeContext(invocation?.runtime);
  }

  static fromRuntimeContext(runtime?: EventRuntimeContext | null): EventIdentity {
    const runId = runtime?.run_id ?? undefined;
    const executionId = runtime?.execution_id ?? undefined;
    const turnId = runtime?.turn_id ?? executionId ?? runId ?? randomUUID();
    const parentEventId = runtime?.parent_event_id ?? runtime?.subagent_lineage?.parent_event_id ?? undefined;
    return new EventIdentity({
      streamId: StreamId.new(randomUUID()),
      conversationId: runtime?.conversation_id ?? undefined,
      runId,
      turnId,
      messageId: runtime?.message_id ?? undefined,
      executionId,
      parentEventId,
    });
  }

  validate(): void {
    StreamId.new(this.streamId.asStr());
    nonEmpty(this.turnId, "turn_id");
    for (const [value, label] of [
      [this.conversationId, "conversation_id"],
      [this.runId, "run_id"],
      [this.messageId, "message_id"],
      [this.executionId, "execution_id"],
      [this.parentEventId, "event_id"],
    ] as const) {
      if (value !== undefined) nonEmpty(value, label);
    }
  }

  withConversationId(value: string): EventIdentity {
    return this.copy({ conversationId: nonEmpty(value, "conversation_id") });
  }

  withRunId(value: string): EventIdentity {
    return this.copy({ runId: nonEmpty(value, "run_id") });
  }

  withMessageId(value: string): EventIdentity {
    return this.copy({ messageId: nonEmpty(value, "message_id") });
  }

  withExecutionId(value: string): EventIdentity {
    return this.copy({ executionId: nonEmpty(value, "execution_id") });
  }

  withParentEventId(value: string): EventIdentity {
    return this.copy({ parentEventId: nonEmpty(value, "event_id") });
  }

  private copy(overrides: Partial<Omit<EventIdentity, "copy" | "validate">>): EventIdentity {
    return new EventIdentity({
      streamId: overrides.streamId ?? this.streamId,
      conversationId: overrides.conversationId ?? this.conversationId,
      runId: overrides.runId ?? this.runId,
      turnId: overrides.turnId ?? this.turnId,
      messageId: overrides.messageId ?? this.messageId,
      executionId: overrides.executionId ?? this.executionId,
      parentEventId: overrides.parentEventId ?? this.parentEventId,
    });
  }
}

File: sdks/python/src/echo_agent_sdk/event_identity_values.py
from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass, replace
from typing import Any
from uuid import uuid4


def _non_empty(value: str, label: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"{label} must not be empty")
    return value


@dataclass(frozen=True)
class StreamId:
    value: str

    def __post_init__(self) -> None:
        object.__setattr__(self, "value", _non_empty(self.value, "event stream_id"))

    @classmethod
    def new(cls, value: str) -> StreamId:
        return cls(value)

    def as_str(self) -> str:
        return self.value

    def __str__(self) -> str:
        return self.value


@dataclass(frozen=True)
class EventId:
    value: str

    def __post_init__(self) -> None:
        object.__setattr__(self, "value", _non_empty(self.value, "event_id"))

    @classmethod
    def new(cls, value: str) -> EventId:
        return cls(value)

    def as_str(self) -> str:
        return self.value

    def __str__(self) -> str:
        return self.value


def _field(value: Any, name: str) -> str | None:
    if isinstance(value, Mapping):
        result = value.get(name)
    else:
        result = getattr(value, name, None)
    return result if result is None or isinstance(result, str) else str(result)


@dataclass(frozen=True)
class EventIdentity:
    stream_id: StreamId
    turn_id: str
    conversation_id: str | None = None
    run_id: str | None = None
    message_id: str | None = None
    execution_id: str | None = None
    parent_event_id: str | None = None

    def __post_init__(self) -> None:
        if not isinstance(self.stream_id, StreamId):
            object.__setattr__(self, "stream_id", StreamId.new(str(self.stream_id)))
        object.__setattr__(self, "turn_id", _non_empty(self.turn_id, "turn_id"))
        for value, label in (
            (self.conversation_id, "conversation_id"),
            (self.run_id, "run_id"),
            (self.message_id, "message_id"),
            (self.execution_id, "execution_id"),
            (self.parent_event_id, "event_id"),
        ):
            if value is not None:
                _non_empty(value, label)

    @classmethod
    def new(cls, stream_id: str, turn_id: str) -> EventIdentity:
        return cls(StreamId.new(stream_id), turn_id)

    @classmethod
    def for_run(cls, run_id: str) -> EventIdentity:
        run_id = _non_empty(run_id, "run_id")
        return cls(
            StreamId.new(str(uuid4())), run_id, run_id=run_id, execution_id=run_id
        )

    @classmethod
    def for_chat(
        cls,
        conversation_id: str | None,
        turn_id: str,
        message_id: str,
        run_id: str | None = None,
    ) -> EventIdentity:
        return cls(
            StreamId.new(str(uuid4())),
            turn_id,
            conversation_id=conversation_id,
            run_id=run_id,
            message_id=_non_empty(message_id, "message_id"),
        )

    @classmethod
    def from_invocation(cls, invocation: Any | None) -> EventIdentity:
        if invocation is None:
            return cls.from_runtime_context(None)
        if isinstance(invocation, Mapping):
            runtime = invocation.get("runtime")
        else:
            runtime = getattr(invocation, "runtime", None)
        return cls.from_runtime_context(runtime)

    @classmethod
    def from_runtime_context(cls, runtime: Any | None) -> EventIdentity:
        run_id = _field(runtime, "run_id")
        execution_id = _field(runtime, "execution_id")
        turn_id = _field(runtime, "turn_id") or execution_id or run_id or str(uuid4())
        lineage = _field(runtime, "subagent_lineage")
        parent_event_id = _field(runtime, "parent_event_id") or _field(
            lineage, "parent_event_id"
        )
        return cls(
            StreamId.new(str(uuid4())),
            turn_id,
            conversation_id=_field(runtime, "conversation_id"),
            run_id=run_id,
            message_id=_field(runtime, "message_id"),
            execution_id=execution_id,
            parent_event_id=parent_event_id,
        )

    def validate(self) -> None:
        StreamId.new(self.stream_id.as_str())
        _non_empty(self.turn_id, "turn_id")

    def with_conversation_id(self, value: str) -> EventIdentity:
        return replace(self, conversation_id=_non_empty(value, "conversation_id"))

    def with_run_id(self, value: str) -> EventIdentity:
        return replace(self, run_id=_non_empty(value, "run_id"))

    def with_message_id(self, value: str) -> EventIdentity:
        return replace(self, message_id=_non_empty(value, "message_id"))

    def with_execution_id(self, value: str) -> EventIdentity:
        return replace(self, execution_id=_non_empty(value, "execution_id"))

    def with_parent_event_id(self, value: str) -> EventIdentity:
        return replace(self, parent_event_id=_non_empty(value, "event_id"))
